File: app/src/main/java/com/drawable/Drawable.java
package com.drawable;

import android.graphics.Canvas;
import android.graphics.Paint;

@SuppressWarnings({"unused", "WeakerAccess"})
public abstract class Drawable {
    private double x;
    private double y;
    private int color;
    private double xVelocity;
    private double yVelocity;

    public Drawable(double x, double y, int color, double xVelocity, double yVelocity) {
        this.x = x;
        this.y = y;
        this.color = color;
        this.xVelocity = xVelocity;
        this.yVelocity = yVelocity;
    }

    public Drawable(double x, double y, int color) {
        this(x, y, color, 0, 0);
    }

    public Drawable(Drawable d) {
        this(d.getX(), d.getY(), d.getColor(), d.getXVelocity(), d.getYVelocity());
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public int getColor() {
        return color;
    }

    public double getXVelocity() {
        return xVelocity;
    }

    public double getYVelocity() {
        return yVelocity;
    }

    public void setX(double x) {
        this.x = x;
    }

    public void setY(double y) {
        this.y = y;
    }

    public void setColor(int color) {
        this.color = color;
    }

    public void setXVelocity(double xVelocity) {
        this.xVelocity = xVelocity;
    }

    public void setYVelocity(double yVelocity) {
        this.yVelocity = yVelocity;
    }

    public void setPos(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public void updatePos(double dx, double dy) {
        this.x += dx;
        this.y += dy;
    }

    protected void animate() {
        if (getX() == 0 && getY() == 0) return;
        setX(getX() + getXVelocity());
        setY(getY() + getYVelocity());
    }

    @Override
    public String toString() {
        return this.getClass().getSimpleName() + ": \n \tX position: " + x + "\n \tY position: " + y +
                "\n \tColor: " + String.format("#%06X", (0xFFFFFF & getColor())) + "\n \t" +
                "X velocity: " + xVelocity + "\n \tY velocity: " + yVelocity;
    }

    public abstract void draw(Canvas canvas, Paint paint);
}

File: app/src/main/java/com/drawable/Drawable.kt
package com.drawable

import android.graphics.Canvas
import android.graphics.Paint

import com.util.Color

@Suppress("unused", "NAME_SHADOWING")
abstract class Drawable
@JvmOverloads constructor(var x: Double, var y: Double, var color: Int, var xVelocity: Double = 0.0, var yVelocity: Double = 0.0) {

    val colorAsHex: String
        get() = String.format("#%06X", 0xFFFFFF and color)

    constructor(x: Double, y: Double, color: Color, xVelocity: Double, yVelocity: Double) : this(x, y, color.int, xVelocity, yVelocity)

    constructor(x: Double, y: Double, color: Color) : this(x, y, color.int, 0.0, 0.0)

    constructor(d: Drawable) : this(d.x, d.y, d.color, d.xVelocity, d.yVelocity)

    fun setColor(color: Color) {
        this.color = color.int
    }

    fun setColor(color: String) {
        var color = color
        if (color.substring(0, 1) != "#") {
            color = "#$color"
        }
        this.color = Color.parseColor(color)
    }

    fun setPos(x: Double, y: Double) {
        this.x = x
        this.y = y
    }

    fun updatePos(dx: Double, dy: Double) {
        this.x += dx
        this.y += dy
    }

    protected open fun animate() {
        x += xVelocity
        y += yVelocity
    }

    override fun toString(): String {
        return this.javaClass.simpleName + ": \n \tX position: " + x + "\n \tY position: " + y +
                "\n \tColor: " + colorAsHex + "\n \t" + "X velocity: " + xVelocity + "\n \tY velocity: " + yVelocity
    }

    abstract fun draw(canvas: Canvas, paint: Paint)
}